### PR TITLE
cacher avoid double locking

### PR DESCRIPTION
**Pull request by p0lyn0mial**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
it turns out that the watcherBookmarkTimeBuckets
is called from only three places/methods: startDispatching, finishDispatching and Watch.
All these methods acquire c.Lock() before touching watcherBookmarkTimeBuckets.

Thus we could remove explicit locking in
watcherBookmarkTimeBuckets since the access is already synced.

xref: https://github.com/kubernetes/enhancements/issues/3157
#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 1]
```diff
@@ -184,7 +184,6 @@ func (i *indexedWatchers) terminateAll(groupResource schema.GroupResource, done
 // second in a bucket, and pop up them once at the timeout. To be more specific,
 // if you set fire time at X, you can get the bookmark within (X-1,X+1) period.
 type watcherBookmarkTimeBuckets struct {
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:186]: it looks like the `watcherBookmarkTimeBuckets` is called from only 3 method: `startDispatching`, `finishDispatching` and `Watch`. All these methods acquire `c.Lock()` before touching `watcherBookmarkTimeBuckets`.

thus we could remove explicit locking in some methods of `watcherBookmarkTimeBuckets` since the access is already synced.

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:186]: I think this would simplify https://github.com/kubernetes/kubernetes/pull/117014 (we could remove the refactoring)

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go [hunk 2]
```diff
@@ -686,7 +686,9 @@ func TestCacherNoLeakWithMultipleWatchers(t *testing.T) {
 			case <-stopCh:
 				return
 			default:
+				cacher.Lock()
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:689]: the drawback is that now low level test need lock before accessing `bookmarkWatchers` but we could change some method names to highlight is is not safe

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go (hunk 1 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:186]: @wojtek-t wdyt?

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:186]: This makes sense to me (especially that other watchers maps are also lock-free and protected by higher-level lock).

So yeah - let's make this happen.

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:186]: done, PTAL.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 3]
```diff
@@ -205,7 +204,7 @@ func newTimeBucketWatchers(clock clock.Clock, bookmarkFrequency time.Duration) *
 
 // adds a watcher to the bucket, if the deadline is before the start, it will be
 // added to the first one.
-func (t *watcherBookmarkTimeBuckets) addWatcher(w *cacheWatcher) bool {
+func (t *watcherBookmarkTimeBuckets) addWatcherThreadUnsafe(w *cacheWatcher) bool {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:207]: Can you also in this line:
https://github.com/kubernetes/kubernetes/blob/e4ee8fded488ee400f03862f83b3d46a3ec8a4c5/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L326

(where `bookmarkWatchers` field is defined)
add an additional comment, that this field when accessed should be protected by the Cacher.lock itself.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go (hunk 2 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:689]: I see you already did that - this looks fine.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go [hunk 4]
```diff
@@ -700,9 +702,9 @@ func TestCacherNoLeakWithMultipleWatchers(t *testing.T) {
 
 	// wait out the expiration period and pop expired watchers
 	time.Sleep(2 * time.Second)
-	cacher.bookmarkWatchers.popExpiredWatchers()
-	cacher.bookmarkWatchers.lock.Lock()
-	defer cacher.bookmarkWatchers.lock.Unlock()
+	cacher.Lock()
+	cacher.bookmarkWatchers.popExpiredWatchersThreadUnsafe()
+	cacher.Unlock()
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:707]: Why did you change this line?
Below we're also accessing bookmarkWatchers, so we should (as previously) do that under lock.

**Review by wojtek-t — COMMENTED** [review]
Couple minor comments - overall LGTM.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go (hunk 3 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:207]: done

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go (hunk 4 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:707]: You didn't understand my comment - we need to aquire a lock.

I'm just saying you shouldn't unlock here, just revert back to "defer unlock"

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:707]: ah, now I see it, `popExpiredWatchersThreadUnsafe` also accesses the `bookmarkWatchers`, good catch, thx!

**Comment by wojtek-t** [comment]
/label tide/merge-method-squash

/lgtm
/approve

Thanks!
